src/libAtomVM/otp_socket.c

```diff
@@ -0,0 +1,1364 @@
+                    term address = socket_tuple_from_addr(ctx, ntohl(addr.sin_addr.s_addr));
+                    term port_number = term_from_int(ntohs(addr.sin_port));
+
+                    term map = term_alloc_map(2, &ctx->heap);
+                    term_set_map_assoc(map, 0, ADDR_ATOM, address);
+                    term_set_map_assoc(map, 1, PORT_ATOM, port_number);
+                    term tuple = port_heap_create_tuple2(&ctx->heap, map, data);
+                    payload = port_heap_create_ok_tuple(&ctx->heap, tuple);
+                } else {
+                    payload = port_heap_create_ok_tuple(&ctx->heap, data);
+                }
+            }
+
+            free(buffer);
+            return payload;
+        }
+    } else {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+}
+
+static term nif_socket_recv_internal(Context *ctx, int argc, term argv[], bool is_recvfrom)
+{
+    if (otp_socket_platform_supports_peek()) {
+        return nif_socket_recv_with_peek(ctx, argc, argv, is_recvfrom);
+    } else {
+        return nif_socket_recv_without_peek(ctx, argc, argv, is_recvfrom);
+    }
+}
+
+static term nif_socket_recv(Context *ctx, int argc, term argv[])
+{
+    TRACE("nif_socket_recv\n");
+    return nif_socket_recv_internal(ctx, argc, argv, false);
+}
+
+static term nif_socket_recvfrom(Context *ctx, int argc, term argv[])
+{
+    TRACE("nif_socket_recvfrom\n");
+    return nif_socket_recv_internal(ctx, argc, argv, true);
+}
+
+//
+// send/sendto
+//
+
+static term nif_socket_send_internal(Context *ctx, int argc, term argv[], bool is_sendto)
+{
+    TRACE("nif_socket_send_internal\n");
+    UNUSED(argc);
+
+    VALIDATE_VALUE(argv[0], term_is_socket);
+
+    GlobalContext *global = ctx->global;
+
+    void *fd_obj_ptr;
+    if (UNLIKELY(!enif_get_resource(erl_nif_env_from_context(ctx), get_socket(argv[0]), socket_resource_type, &fd_obj_ptr))) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+    struct SocketFd *fd_obj = (struct SocketFd *) fd_obj_ptr;
+    if (fd_obj->fd) {
+
+        term data = argv[1];
+        term dest = term_invalid_term();
+        if (is_sendto) {
+            dest = argv[2];
+        }
+
+        // TODO make non-blocking
+
+        char *buf = NULL;
+        size_t len = 0;
+        if (term_is_binary(data)) {
+            buf = (char *) term_binary_data(data);
+            len = term_binary_size(data);
+        } else if (term_is_list(data)) {
+            InteropFunctionResult result = interop_iolist_size(data, &len);
+            if (UNLIKELY(result == InteropMemoryAllocFail)) {
+                AVM_LOGW(TAG, "Failed to allocate memory: %s:%i.\n", __FILE__, __LINE__);
+                RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+            }
+            if (UNLIKELY(result == InteropBadArg)) {
+                TRACE("badarg: input data is not an iolist\n");
+                RAISE_ERROR(BADARG_ATOM);
+            }
+            // TODO: send an iolist without allocating a buffer, maybe by iterative sends
```

Comment:
Alternatively convert to a binary first, as OTP puts it we need to call send as little as possible, especially for UDP where it must be only one packet (even TCP, the semantic can be a little different), so we have to copy all data to a single buffer.

Reply:
Now done in erlang